spock-core/src/main/java/org/spockframework/runtime/extension/builtin/RetryIterationInterceptor.java

```diff
@@ -74,26 +77,32 @@ public void intercept(IMethodInvocation invocation) throws Throwable {
   }
 
   private class IterationState {
-    private final ThreadLocal<Boolean> finalIteration = ThreadLocal.withInitial(() -> false);
-    private final ThreadLocal<List<Throwable>> throwables = new ThreadLocal<>();
-
-    void startIteration() {
-      throwables.set(new ArrayList<>());
+    private final Map<IterationInfo, AtomicBoolean> finalIteration = new ConcurrentHashMap<>();
+    private final Map<IterationInfo, List<Throwable>> throwables = new ConcurrentHashMap<>();
```

Comment:
I would prefer the type `ConcurrentMap<IterationInfo, CopyOnWriteArrayList<Throwable>>`.
This would cearly state that the field is threadsafe.